native/wasmex/src/memory.rs

```diff
@@ -209,27 +209,28 @@ pub fn read_binary<'a>(env: Env<'a>, args: &[Term<'a>]) -> Result<Term<'a>, Erro
 
     if end > view.len() {
         return Err(Error::RaiseTerm(Box::new(
-            "Out of bound: The given binary will write out of memory",
+            "Out of bound: The given binary will read out of memory",
         )));
     }
 
-    let mut binary: Vec<u8> = Vec::new();
+    let mut binary: OwnedBinary = OwnedBinary::new(length).unwrap();
+
     for i in start..end {
         let value = view[i].get();
-        binary.push(value);
-        if value == 0 {
-            break;
-        }
+        binary[i - start] = value;
     }
-    Ok(binary.encode(env))
+
+    let final_binary: Binary = Binary::from_owned(binary, env);
+
+    Ok(final_binary.encode(env))
```

Comment:
What about this instead?

* We could shadow the `binary` variable (as it serves the same purpose) from being the `OwnedBinary` to the normal `Binary`.
* we use higher-level rust to copy the content, instead of a manual loop
* I tried it locally and the tests work for me

```rust
    let mut binary: OwnedBinary = OwnedBinary::new(length).unwrap();
    let data = view[start..end]
        .into_iter()
        .map(|cell| cell.get())
        .collect::<Vec<u8>>();
    binary.copy_from_slice(&data);

    let binary: Binary = Binary::from_owned(binary, env);

    Ok(binary.encode(env))
```

Reply:
Makes sense to me 👍